chain/db/name.go

```diff
@@ -7,96 +7,153 @@ import (
 		return err
 	}
 
+	delete(sdb.names, registrantId)
+	delete(sdb.nameRegistrants, nameId)
+
 	return nil
 }
 
-func (cs *ChainStore) GetName(registrant []byte) (*string, error) {
-	registrantKey := generateRegistrantKey(registrant)
-
-	data, err := cs.st.Get(registrantKey)
+func (sdb *StateDB) deleteNameForRegistrant(registrant []byte) error {
+	name, err := sdb.getName(registrant)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	if name == "" {
+		return nil
 	}
 
-	r := bytes.NewReader(data)
+	registrantId := getRegistrantId(registrant)
+	nameId := getNameId(name)
 
-	name, err := serialization.ReadVarString(r)
-	if err != nil {
-		return nil, err
-	}
+	sdb.names[registrantId] = ""
+	sdb.nameRegistrants[nameId] = nil
 
-	return &name, nil
+	return nil
 }
 
-func (cs *ChainStore) GetRegistrant(name string) ([]byte, error) {
-	nameKey := generateNameKey(name)
+func (cs *ChainStore) DeleteName(registrant []byte) error {
+	return cs.States.deleteNameForRegistrant(registrant)
+}
 
-	data, err := cs.st.Get(nameKey)
-	if err != nil {
-		return nil, err
+func (sdb *StateDB) getName(registrant []byte) (string, error) {
+	registrantId := getRegistrantId(registrant)
+	var name string
+	var ok bool
+	if name, ok = sdb.names[registrantId]; !ok {
+		enc, err := sdb.trie.TryGet(append(NameRegistrantPrefix, registrantId...))
+		if err != nil {
+			return "", err
+		}
+
+		if len(enc) > 0 {
+			buff := bytes.NewBuffer(enc)
+			name, err = serialization.ReadVarString(buff)
+			if err != nil {
+				return "", err
+			}
+
+			nameId := getNameId(name)
+			sdb.names[registrantId] = name
+			sdb.nameRegistrants[nameId] = registrant
+		}
 	}
 
-	r := bytes.NewReader(data)
+	return name, nil
+}
 
-	registrant, err := serialization.ReadVarBytes(r)
-	if err != nil {
-		return nil, err
+func (sdb *StateDB) getRegistrant(name string) ([]byte, error) {
+	nameId := getNameId(name)
+	var registrant []byte
+	var ok bool
+	if registrant, ok = sdb.nameRegistrants[nameId]; !ok {
+		enc, err := sdb.trie.TryGet(append(NamePrefix, nameId...))
+		if err != nil {
+			return nil, err
+		}
+
+		if len(enc) > 0 {
+			buff := bytes.NewBuffer(enc)
+			registrant, err = serialization.ReadVarBytes(buff)
+			if err != nil {
+				return nil, err
+			}
+
+			registrantId := getRegistrantId(registrant)
+			sdb.names[registrantId] = name
+			sdb.nameRegistrants[nameId] = registrant
+		}
 	}
 
 	return registrant, nil
 }
+
+func (cs *ChainStore) GetName(registrant []byte) (string, error) {
+	return cs.States.getName(registrant)
+}
+
+func (cs *ChainStore) GetRegistrant(name string) ([]byte, error) {
+	return cs.States.getRegistrant(name)
+}
+
+func (sdb *StateDB) FinalizeNames(commit bool) {
+	for registrantId, name := range sdb.names {
+		nameId := getNameId(name)
+		registrant := sdb.nameRegistrants[nameId]
+		if name == "" {
+			sdb.deleteName(registrantId, nameId)
+		} else {
+			sdb.updateName(registrant, name)
+		}
+		if commit {
+			delete(sdb.names, registrantId)
```

Comment:
Why we need to remove these when `name != ""` ?

Reply:
same logic as with account and nanopay. basically if we commit then we clean up.

Reply:
But I don't see FinalizeNames being called anywhere?

Reply:
it's called in statedb.go/Finalize